workspaces/arborist/lib/arborist/isolated-reifier.js

```diff
@@ -0,0 +1,449 @@
+   *
+   * Since the RFC requires leaving the package-lock in place, this approach temporarily replaces
+   * the tree state for a couple of steps of reifying.
+   *
+   **/
+  async [_makeIdealGraph] (options) {
+    /* Make sure that the ideal tree is build as the rest of
+     * the algorithm depends on it.
+     */
+    const bitOpt = {
+      ...options,
+      complete: false,
+    }
+    await this.buildIdealTree(bitOpt)
+    const idealTree = this.idealTree
+
+    function memoize (fn) {
+      const memo = new Map()
+      return async function (arg) {
+        const key = arg
+        if (memo.has(key)) {
+          return memo.get(key)
+        }
+        const result = {}
+        memo.set(key, result)
+        await fn(result, arg)
+        return result
+      }
+    }
+
+    this.rootNode = {}
+    const root = this.rootNode
+    this.counter = 0
+
+    // memoize to cache generating proxy Nodes
+    this.externalProxyMemo = memoize(this.externalProxy.bind(this))
+    this.workspaceProxyMemo = memoize(this.workspaceProxy.bind(this))
+
+    root.external = []
+    root.isProjectRoot = true
+    root.localLocation = idealTree.location
+    root.localPath = idealTree.path
+    root.workspaces = await Promise.all([...idealTree.fsChildren.values()].map(this.workspaceProxyMemo))
+    const processed = new Set()
+    const queue = [idealTree, ...idealTree.fsChildren]
+    while (queue.length !== 0) {
+      const next = queue.pop()
+      if (processed.has(next.location)) {
+        continue
+      }
+      processed.add(next.location)
+      next.edgesOut.forEach(e => {
+        if (!e.to || (next.package.bundleDependencies || next.package.bundledDependencies || []).includes(e.to.name)) {
+          return
+        }
+        queue.push(e.to)
+      })
+      if (!next.isProjectRoot && !next.isWorkspace) {
+        root.external.push(await this.externalProxyMemo(next))
+      }
+    }
+
+    await this.assignCommonProperties(idealTree, root)
+
+    this.idealGraph = root
+  }
+
+  async workspaceProxy (result, node) {
+    result.localLocation = node.location
+    result.localPath = node.path
+    result.isWorkspace = true
+    result.resolved = node.resolved
+    await this.assignCommonProperties(node, result)
+  }
+
+  async externalProxy (result, node) {
+    await this.assignCommonProperties(node, result)
+    if (node.hasShrinkwrap) {
+      const dir = join(
+        node.root.path,
+        'node_modules',
+        '.store',
+        `${node.name}@${node.version}`
+      )
+      fs.mkdirSync(dir, { recursive: true })
+      // TODO this approach feels wrong
+      // and shouldn't be necessary for shrinkwraps
+      await pacote.extract(node.resolved, dir, {
+        ...this.options,
+        resolved: node.resolved,
+        integrity: node.integrity,
+      })
+      const Arborist = this.constructor
+      const arb = new Arborist({ ...this.options, path: dir })
+      await arb[_makeIdealGraph]({ dev: false })
```

Comment:
Why does this need to use a quasi-private method instead of using buildIdealTree?

Reply:
Said private method does run buildIdealTree and starts the first steps of getting the new tree ready for conversion.

Reply:
What steps are these that non-npm consumers of arborist might want to use? :-)